cms/djangoapps/contentstore/signals/course_dates.py

```diff
@@ -0,0 +1,79 @@
+"""Signal handlers for writing course dates into edx_when."""
+from __future__ import absolute_import, unicode_literals
+
+import logging
+
+from django.dispatch import receiver
+from six import text_type
+from xblock.fields import Scope
+from xmodule.modulestore.django import SignalHandler, modulestore
+
+from edx_when.api import FIELDS_TO_EXTRACT, set_dates_for_course
+
+log = logging.getLogger(__name__)
+
+
+def date_field_values(date_fields, xblock):
+    """
+    Read field values for the specified date fields from the supplied xblock.
+    """
+    result = {}
+    for field_name in date_fields:
+        if field_name not in xblock.fields:
+            continue
+        field = xblock.fields[field_name]
+        if field.scope == Scope.settings and field.is_set_on(xblock):
+            try:
+                result[field.name] = field.read_from(xblock)
+            except TypeError as exception:
+                exception_message = "{message}, Block-location:{location}, Field-name:{field_name}".format(
+                    message=text_type(exception),
+                    location=text_type(xblock.location),
+                    field_name=field.name
+                )
+                raise TypeError(exception_message)
+    return result
+
+
+def extract_dates_from_course(course):
+    """
+    Extract all dates from the supplied course.
+    """
+    from xmodule.modulestore.django import modulestore  # pylint: disable=import-error
+
+    log.info('Publishing course dates for %s', course.id)
+    if course.self_paced:
+        metadata = date_field_values(FIELDS_TO_EXTRACT, course)
+        # self-paced courses may accidentally have a course due date
+        metadata.pop('due', None)
+        date_items = [(course.location, metadata)]
+    else:
+        date_items = []
+        items = modulestore().get_items(course.id)
+        log.info('extracting dates from %d items in %s', len(items), course.id)
+        for item in items:
+            date_items.append((item.location, date_field_values(FIELDS_TO_EXTRACT, item)))
+    return date_items
+
+
+@receiver(SignalHandler.course_published)
+def extract_dates(sender, course_key, **kwargs):  # pylint: disable=unused-argument
```

Comment:
Is there already a signal handler for course_published that exists?  would it make sense to have this logic in that one so that future devs only have to look in that one place?  just a thought

Reply:
I prefer to keep the particular logic that's happening close together, rather than just bundling it on what signal it's listening to. That way, this function is specific to dates, and the other signal handling can be specific to the other tasks that need to happen on course publish.